cmd/csi-addons/replication.go

```diff
@@ -59,28 +60,67 @@ func (rep *VolumeReplicationBase) Init(c *command) error {
 		return errors.New("secret name is not set")
 	}
 
-	pv, err := getKubernetesClient().CoreV1().PersistentVolumes().Get(context.Background(), c.persistentVolume, metav1.GetOptions{})
-	if err != nil {
-		return fmt.Errorf("failed to get pv %q", c.persistentVolume)
-	}
-
-	if pv.Spec.CSI == nil {
-		return fmt.Errorf("pv %q is not a CSI volume", c.persistentVolume)
-	}
-
-	if pv.Spec.CSI.VolumeHandle == "" {
-		return errors.New("volume ID is not set")
+	if c.persistentVolume != "" {
+		pv, err := getKubernetesClient().CoreV1().PersistentVolumes().Get(context.Background(), c.persistentVolume, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get pv %q", c.persistentVolume)
+		}
+
+		if pv.Spec.CSI == nil {
+			return fmt.Errorf("pv %q is not a CSI volume", c.persistentVolume)
+		}
+
+		if pv.Spec.CSI.VolumeHandle == "" {
+			return errors.New("volume ID is not set")
+		}
+		rep.volumeID = pv.Spec.CSI.VolumeHandle
+		return nil
+	} else if c.volumeGroupReplicationContent != "" {
+		vgrc, err := getVolumeReplicationClient().getVolumeGroupReplicationContent(context.Background(), c.volumeGroupReplicationContent, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get VolumeGroupReplicationContent %q", c.volumeGroupReplicationContent)
+		}
+		if vgrc.Spec.VolumeGroupReplicationHandle == "" {
+			return errors.New("volume group ID is not set")
+		}
+		rep.groupID = vgrc.Spec.VolumeGroupReplicationHandle
+		return nil
 	}
-	rep.volumeID = pv.Spec.CSI.VolumeHandle
 
-	return nil
+	return errors.New("only one of persistentVolume or volumeGroupReplicationContent should be set")
 }
 
 // EnableVolumeReplication executes the EnableVolumeReplication operation.
 type EnableVolumeReplication struct {
 	VolumeReplicationBase
 }
 
+func (v VolumeReplicationBase) setReplicationSource(req *proto.ReplicationSource) error {
+	switch {
+	case req == nil:
+		return errors.New("replication source is not set")
+	case v.volumeID != "" && v.groupID != "":
+		return errors.New("only one of volumeID or groupID should be set")
+	case v.volumeID != "":
+		req.Type = &proto.ReplicationSource_Volume{
+			Volume: &proto.ReplicationSource_VolumeSource{
+				VolumeId: v.volumeID,
+			},
+		}
+		return nil
+	case v.groupID != "":
+		req = &proto.ReplicationSource{
```

Comment:
this does not look correct, you don't modify the passed `req` here.

Reply:
@nixpanic This is the replication Spec which need to be set nothing else is modified here.

Reply:
But is it intended to overwrite the `req` variable this way? The new `req` is not returned, so the caller of the function will never see it. I expect a warning from golangci about this too.

The original (`req.Type`) is not updated as is done on line 105.

Reply:
Sorry, its an overlook from my side, you are correct we should not overwrite the req rather set the .Type as its done in other place.